packages/ui/src/server/controllers/vcs.ts

```diff
@@ -0,0 +1,261 @@
+    UNAUTHORIZED:
+        'The service cannot authorize you. Check the token. You can add new token in the section Settings -> VCS',
+};
+
+const cookieNameByVcsId = (vcsId: string) => `vcs_${vcsId}`;
+const sendApiError = (res: Response, error: AxiosError) => {
+    let status = error.response?.status;
+    let message = error.response?.statusText;
+
+    if (status === 401) {
+        status = 500;
+        message = ERROR_MESSAGE.UNAUTHORIZED;
+    }
+
+    sendError(res, message || error.message, status || 500);
+};
+
+export const getVcsConfig: AppRouteHandler = (req, res) => {
+    const cookies = req.cookies;
+    try {
+        const config = VcsFactory.getVcsSettings(req);
+        if (!config.length) {
+            sendError(res, ERROR_MESSAGE.VCS_CONFIG_NOT_FOUND, 404);
+            return;
+        }
+
+        const response = config.map<VcsConfig>((item) => ({
+            ...item,
+            token: cookieNameByVcsId(item.id) in cookies,
+        }));
+        res.status(200).json(response);
+    } catch (e) {
+        req.ctx.logError('Error getting vcs configuration', e);
+        sendError(res, e, 500);
+    }
+};
+
+export const createToken: AppRouteHandler = (req, res) => {
+    const {vcsId, token} = req.body;
+
+    if (!vcsId) {
+        sendError(res, ERROR_MESSAGE.VCS_ID_REQUIRED, 400);
+        return;
+    }
+
+    if (!token) {
+        sendError(res, ERROR_MESSAGE.TOKEN_REQUIRED, 400);
+        return;
+    }
+
+    try {
+        const vcs = VcsFactory.getVcsSettingById(req, vcsId);
+        res.cookie(cookieNameByVcsId(vcs.id), token, COOKIE_OPTIONS);
+        res.status(200).json({message: `${cookieNameByVcsId(vcs.id)} token created`});
+    } catch (e) {
+        req.ctx.logError('Error create token', e);
+        sendError(res, e, 500);
+    }
+};
+
+export const removeToken: AppRouteHandler = (req, res) => {
+    const {vcsId} = req.body;
+
+    if (!vcsId) {
+        sendError(res, ERROR_MESSAGE.VCS_ID_REQUIRED, 400);
+    }
+
+    res.clearCookie(cookieNameByVcsId(vcsId));
+    res.status(200).json({success: true});
+};
+
+const getProjectIdByRepository = (repository: VcsRepository) => {
+    return 'projectId' in repository
+        ? repository.projectId
+        : `${repository.owner}/${repository.name}`;
+};
+
+export const getRepositories: AppRouteHandler = async (req, res) => {
+    const cookies = req.cookies;
+    const vcsId = req.query.vcsId as string | undefined;
+
+    if (!vcsId) {
+        sendError(res, ERROR_MESSAGE.VCS_ID_REQUIRED, 400);
+        return;
+    }
+
+    const token = cookies[cookieNameByVcsId(vcsId)];
+    if (!token) {
+        sendError(res, ERROR_MESSAGE.TOKEN_REQUIRED, 400);
+        return;
+    }
+
+    try {
+        const vcs = VcsFactory.getVcsSettingById(req, vcsId);
+        const api = VcsFactory.createVcs(vcs, token);
+        const response = await api.getRepositories();
+        if (response instanceof AxiosError) {
```

Comment:
Here and below methods of `api` never return instances of AxiosError. Such check should be mentioned once in sendApiError.